Sermons list pagination

avorg opened the pull request.

narthur committed:
Generate pages of results

narthur committed:
start using lodash

narthur committed:
pass pagination info to template

narthur committed:
Build and style sermons pagination

narthur committed:
Update _redirects

narthur committed:
fix two vulnerabilities

narthur committed:
split ci into two jobs

mattleff commented on src/templates/sermons.css.map:
```diff
@@ -1 +1 @@
-{"version":3,"sourceRoot":"","sources":[],"names":[],"mappings":"","file":"sermons.css"}
\ No newline at end of file
+{"version":3,"sourceRoot":"","sources":["sermons.scss"],"names":[],"mappings":"AACE;EACE;EACA;EACA;;AACA;EACE;EACA;;AAEA;EACE;EACA;EACA;EACA;EACA;;AAGF;EACE;;AAIJ;EACE;EACA;EACA","file":"sermons.css"}
```
Do we want/need the `.css.map` committed? Is it possible to generate this in the build step and not have it in version control? I guess that applies to the `.css` too since it's coming from a `.scss` source file.

narthur replied:
Currently I'm using an IDE watcher to do the transcompilation, there's nothing configured in Gatsby or Netlify that would do it for me later.

narthur replied:
Huh, who knew?

https://www.gatsbyjs.org/docs/sass/

mattleff commented on src/templates/sermons.factory.js:
```diff
@@ -1,14 +1,20 @@
 {
     avorg {
-        sermons(language:${key}) {
+        sermons(language:${langKey}, after:"${cursor}") {
```
Rather than using string interpolation a better practice is to use query variables: https://www.gatsbyjs.org/docs/graphql-reference/#query-variables.

narthur replied:
Reference: https://github.com/thalseth/kystleik/blob/32691a7e8b5700deb57f15985291d8e47e5a67e5/web/gatsby-node.js#L73-L91

mattleff commented on src/templates/sermons.factory.js:
```diff
@@ -25,16 +31,42 @@ exports.createPages = async (graphql, createPage) => {
-            context: {nodes}
-        })
+
+    const sermons = _.get(result, 'data.avorg.sermons'),
```
👍

mattleff commented on src/templates/sermons.factory.js:
```diff
@@ -25,16 +31,42 @@ exports.createPages = async (graphql, createPage) => {
+    const langKeys = Object.keys(constants.languages)
+
+    await Promise.all(langKeys.map(async (key) => {
+        return await createPagesByLang(key, graphql, createPage)
```
You either want `return` or `await` here, not both. I would do it like this:

```js
await Promise.all(langKeys.map((key) => createPagesByLang(key, graphql, createPage)));
```

narthur replied:
Haha, I wondered why that change had absolutely no effect on build times

mattleff commented on src/templates/sermons.js:
```diff
@@ -3,12 +3,30 @@ import Layout from "../components/layout"
-    const sermons = pageContext.nodes;
+    const sermons = pageContext.nodes,
+        {total, current} = pageContext.pagination,
+        pageStart = Math.max(1, current - 10),
```
We can nix the magic number here if we give it a name, something like `showAdjacentPages = 10`. That helps indicate what the `10`s on lines 8 + 9 are doing.

mattleff commented on src/templates/sermons.js:
```diff
@@ -3,12 +3,30 @@ import Layout from "../components/layout"
+        pagePrevious = Math.max(1, current - 1),
+        pageNext = Math.min(total, current + 1);
+
+    let pageLinks = []
```
Since you're producing an array of page links from a range of numbers it might be simpler to use `_.range()` and `.map()`:

```js
const pageLinks = _.range(pageStart, pageEnd + 1).map((i) => 
    <li className={current === i ? "active" : null}>
        <a href={`/en/sermons/page/${i}`}>{i}</a>
    </li>);
```

mattleff commented on src/templates/sermons.js:
```diff
@@ -3,12 +3,30 @@ import Layout from "../components/layout"
+            <ul>{sermons.map((n, i) => <li>{n.title}</li>)}</ul>
+
+            <ul className={'template-sermons__pagination'}>
+                <li><a href={`/en/sermons/page/${pagePrevious}`}>{'<'}</a></li>
```
Should this link be disabled if there isn't a previous page?

mattleff commented on src/templates/sermons.js:
```diff
@@ -3,12 +3,30 @@ import Layout from "../components/layout"
+
+            <ul className={'template-sermons__pagination'}>
+                <li><a href={`/en/sermons/page/${pagePrevious}`}>{'<'}</a></li>
+                <li><a href={`/en/sermons/page/1`}>First</a></li>
```
Should this link always be shown, even next to a `1` link?

mattleff commented on src/templates/sermons.scss:
```diff
@@ -0,0 +1,30 @@
+      margin-right: .2em;
+
+      a {
+        color: #747474;
```
It would be good to start extracting these colors into a `global.colors.scss` or something like that. Then when we want to start tweaking/reconciling color variations we won't have as many places to change.

mattleff commented on src/templates/sermons.spec.js:
```diff
@@ -6,60 +6,63 @@ jest.mock(`path`)
     it("passes sermons to createPage", async () => {
-        const sermons = ["sermons"]
+        const sermons = ["sermons"],
+            returnValue = {data:{avorg:{sermons:{nodes:sermons}}}}
```
This isn't too bad, but you could consider using `_.set()`.

mattleff commented on static/_redirects:
```diff
@@ -1,7 +1,23 @@
+/ja/sermons/:id/:slug   /ja/sermons/:id     200
+/ru/sermons/:id/:slug   /ru/sermons/:id     200
+
+# Legacy
```
👍 Good separation

mattleff reviewed (approved):
Great work here! Just a few suggestions to consider. 👍

narthur committed:
Attempt ci fix

narthur committed:
stop async'ing promise.all

Has basically no impact on build performance (1042 >> 1088 seconds during one test locally)

narthur committed:
Extract pagination component

narthur committed:
use gatsby for sass compilation; start work on dev config

narthur committed:
revert filesystem changes

narthur committed:
Rename sermon factory tests file

narthur committed:
Extract queries.getPages

narthur committed:
Refactor sermons factory

narthur committed:
limit page quantity in dev mode

narthur committed:
ensure sermon data gets passed to createPage

narthur committed:
Refactor sermon factory tests

narthur committed:
fix sermon url generation

narthur committed:
fix security vulnerability

narthur committed:
Switch to using query variables

narthur committed:
Fix unit test

narthur committed:
extract common styling

mattleff commented on gatsby-config.js:
```diff
@@ -1,3 +1,7 @@
+const constants = require(`./src/constants`)
+
+console.log(`env: ${constants.environment}`)
```
Nix?

narthur replied:
Going to leave this one here for now. Just adds a log message during the build process. :)

mattleff reviewed (approved):
Great work! :shipit:
